pkg/storage/cloudimpl/aws_kms.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+package cloudimpl
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/cockroachdb/cockroach/pkg/storage/cloud"
+	"github.com/cockroachdb/errors"
+)
+
+const awsScheme = "aws"
+
+type awsKMS struct {
+	kms                 *kms.KMS
+	customerMasterKeyID string
+}
+
+var _ cloud.KMS = &awsKMS{}
+
+func init() {
+	cloud.RegisterKMSFromURIFactory(MakeKMSFromURI, awsScheme)
+}
+
+type kmsURIParams struct {
+	accessKey string
+	secret    string
+	tempToken string
+	endpoint  string
+	region    string
+	auth      string
+}
+
+func resolveKMSURIParams(kmsURI url.URL) kmsURIParams {
+	kmsURIParams := kmsURIParams{
+		kmsURI.Query().Get(AWSAccessKeyParam),
```

Comment:
I've never said "I'm glad I skipped those struct field names" but I sure have said the opposite.